Console command

logicminds opened the pull request:
Adds a new command called `console` that invokes the puppet debugger console from within a module.   This command respects the puppet-version, puppet-dev, and pe-version options but has to parse them manually due to how CRI works and becoming a broker to the puppet debugger command.  Otherwise the `pdk console` command works just like the puppet debugger command and can be passed the same exact options.  In fact using `pdk console -h` will actually return the help info from the puppet debugger command.  This is the reason why I had to skip option parsing. 

At the moment this only works inside a module due to some limitations of the PDK and external gems.  However, working inside a module allows for a scoped invocation of the puppet debugger scoped for the module.   This allows for users to easily test out their module code, facts, datatypes and more without having to supply the modulepath.

```
pdk console
pdk (INFO): Console can only be run from inside a valid module directory
```

This also relies on the user having the puppet-debugger gem in their Gemfile.  The following [PR](https://github.com/puppetlabs/puppet-module-gems/pull/117) should ensure the user has this already.  Otherwise the user would need to add puppet-debugger to the .sync.yml file.  If the user does not have the puppet-debugger gem they will get the following error.  Once solution to work around this is to figure out if they have the gem present already, but this procedure slows down the console command and was removed from the PR.


```
pdk console
Error: Unknown Puppet subcommand 'debugger'
See 'puppet help' for help on available puppet subcommands
```

This PR also relies on #753 and #755 and I have included them as separate PRs and also in this branch.

The demo below shows several invocations with different puppet versions which is quite handy.  The delay you may notice is `PDK::Util::Bundler.ensure_bundle!(puppet_env[:gemset])` being run to ensure the correct puppet version is installed.


![pdk console demo](https://user-images.githubusercontent.com/563900/65357887-0cacff00-dbad-11e9-8b73-df4fa3fce006.gif)

glennsarti commented:
@logicminds  I assume this relates to https://github.com/puppetlabs/pdk-planning/pull/44

May need to update the RFC with the new name of  `console`

glennsarti commented on README.md:
```diff
@@ -105,6 +105,20 @@ This command runs all available unit tests.
 
 ## Experimental features
 
+### `pdk console` command
```
@jbondpdx Can you give these docs a quick look.

* `To use just execute` avoid `just` and `easy`

* ``~/modules/ntp  $ ... ` Avoid shell specific things e.g. a PowerShell user would see this prefixed differently.

logicminds replied:
I update the README with the fixes and also added some extra bits I thought were important.

jbondpdx replied:
Sorry I didn't see this sooner, @glennsarti , @logicminds ... I think it got lost in a mass of GitHub emails. Thank you for both for these docs!

logicminds commented:
Added another command that helps the user if the puppet-debugger gem is missing.

```
pdk console
pdk (FATAL): Your Gemfile is missing the puppet-debugger gem.  You can add the missing gem
by updating your /Users/cosman/blockops/sfmg/crossbelt_data_sfmg/.sync.yml file with the following
and running pdk update.

Gemfile:
  required:
    ":development":
      - gem: puppet-debugger
        version: "~> 0.14"
```

logicminds commented:
@rodjek @glennsarti  I spent some time refactoring my tests and they no longer run the debugger and instead just validate the args being passed through to the debugger.  We know the debugger works so there is no need to test it's functionality.  Not running the debugger speeds up the tests greatly and removes the need for the dependency as well.   This strategy might be something we can do with other pdk commands as well. 

Anyways, this is ready to go.

glennsarti commented on Gemfile:
```diff
@@ -15,6 +15,9 @@ group :development do
   gem 'pry-byebug', '~> 3.4'
+  # if you want to test out the pdk console you will need puppet-debugger gem
+  # the unit tests don't actually call the debugger, so this is not needed
+  # gem 'puppet-debugger' # only needed in bundler sandbox
```
Won't stop merging.  @rodjek , is the pattern for "optional" gems to guard them behind an environment variable, or prompt users to use the `Gemfile.local` file?

logicminds replied:
If left enabled the license checker fails

logicminds replied:
If left enabled the license checker fails

logicminds replied:
I am just removing this in latest push.

glennsarti commented on README.md:
````diff
@@ -105,6 +106,46 @@ This command runs all available unit tests.
+```
+pdk console -h
+Error: Unknown Puppet subcommand 'debugger'
+See 'puppet help' for help on available puppet subcommands
````
`See 'puppet help' for help on available puppet subcommands` seems the wrong command here.  Should be `pdk help`?

logicminds replied:
Because option parsing is disabled there is not a good way to intercept the help command as parsing is either on or off.  As a result the help or -h is sent to the puppet debugger face.  However, because the debugger is not installed on this system you get that response.  This error should be caught.  Assume you have the latest code?

logicminds replied:
I did force push almost 20 times so you might be running older code.

```
pdk help console
NAME
    console - (Experimental) Start a session of the puppet debugger console.

USAGE
    pdk console [console_options]

DESCRIPTION
    The pdk console runs a interactive session of the puppet debugger tool to
    test out snippets of code, run language evaluations, datatype prototyping
    and much more.  A virtual playground for your puppet code! For usage
    details see the puppet debugger docs at https://docs.puppet-debugger.com.

OPTIONS
       --pe-version=<value>          Puppet Enterprise version to run tests
                                     or validations against.
       --puppet-dev                  When specified, PDK will validate or
                                     test against the current Puppet source
                                     from github.com. To use this option, you
                                     must have network access to
                                     https://github.com.
       --puppet-version=<value>      Puppet version to run tests or
                                     validations against.

OPTIONS FOR PDK
    -d --debug                       Enable debug output.
    -f --format=<value>              Specify desired output format. Valid
                                     formats are 'junit', 'text'. You may
                                     also specify a file to which the
                                     formatted output is sent, for example:
                                     '--format=junit:report.xml'. This option
                                     may be specified multiple times if each
                                     option specifies a distinct target file.
    -h --help                        Show help for this command.
       --version                     Show version of pdk.
 ~/nwops/gofish   master ●  pdk console -h
pdk (FATAL): Your Gemfile is missing the puppet-debugger gem.  You can add the missing gem
by updating your /Users/cosman/nwops/gofish/.sync.yml file with the following
and running pdk update.

Gemfile:
  required:
    ":development":
      - gem: puppet-debugger
        version: "~> 0.14"
```

glennsarti commented on lib/pdk/cli/console.rb:
```diff
@@ -0,0 +1,126 @@
+
+      PDK::CLI::Util.analytics_screen_view('console', args)
+
+      # TODO: figure out if we need to remove default configs set by puppet
```
Should TODO still be here?

logicminds replied:
Yes.  I think the args should be merged

logicminds replied:
Yes.  I think the args should be merged

logicminds replied:
Going to fix this in the next iteration. I really wanted to see how people would use it and get some feedback.

glennsarti commented on spec/unit/pdk/cli/console_spec.rb:
```diff
@@ -0,0 +1,174 @@
+      allow(PDK::Util).to receive(:module_root).and_return('/modules/ntp')
+      args = [
+        PDK::CLI::Exec.bundle_bin, 'exec', 'puppet', 'debugger', '--run-once', '--quiet', '--execute=$foo = 123',
+        '--basemodulepath=/modules/ntp/spec/fixtures/modules:/modules/ntp/modules',
```
Worth testing here with Windows style paths e.g.
`--basemodulepath=C:\modules\ntp\spec\fixtures\modules;C:\modules\ntp\modules`

logicminds replied:
Added a new test to cover the windows bit.  Also added some helpers to identify what platform.

glennsarti commented:
Hrmm the `environmentpath` is breaking on my laptop

```
pdk (DEBUG): Executing 'bundle.bat exec puppet debugger --basemodulepath=C:/Source/tmp/pdkmodule/spec/fixtures/modules:C:/Source/tmp/pdkmodule/modules --environmentpath=C:/Source/tmp/pdkmodule/spec/fixtures/modules:C:/Source/tmp/pdkmodule/modules --modulepath=C:/Source/tmp/pdkmodule/spec/fixtures/modules:C:/Source/tmp/pdkmodule/modules' interactively
Traceback (most recent call last):
        7: from C:/Users/glenn.sarti/AppData/Local/PDK/cache/ruby/2.5.0/bin/puppet:23:in `<main>'
        6: from C:/Users/glenn.sarti/AppData/Local/PDK/cache/ruby/2.5.0/bin/puppet:23:in `load'
        5: from C:/Users/glenn.sarti/AppData/Local/PDK/cache/ruby/2.5.0/gems/puppet-6.10.0-x64-mingw32/bin/puppet:5:in `<top (required)>'
        4: from C:/Users/glenn.sarti/AppData/Local/PDK/cache/ruby/2.5.0/gems/puppet-6.10.0-x64-mingw32/lib/puppet/util/command_line.rb:77:in `execute'
        3: from C:/Users/glenn.sarti/AppData/Local/PDK/cache/ruby/2.5.0/gems/puppet-6.10.0-x64-mingw32/lib/puppet/util/command_line.rb:143:in `run'
        2: from C:/Users/glenn.sarti/AppData/Local/PDK/cache/ruby/2.5.0/gems/puppet-6.10.0-x64-mingw32/lib/puppet/application.rb:370:in `run'
        1: from C:/Users/glenn.sarti/AppData/Local/PDK/cache/ruby/2.5.0/gems/puppet-6.10.0-x64-mingw32/lib/puppet/application_support.rb:34:in `push_application_context'
C:/Users/glenn.sarti/AppData/Local/PDK/cache/ruby/2.5.0/gems/puppet-6.10.0-x64-mingw32/lib/puppet/environments.rb:38:in `get!': Could not find a directory environment named 'workstation_production' anywhere in the path: C:/Source/tmp/pdkmodule/spec/fixtures/modules:C:/Source/tmp/pdkmodule/modules. Does the directory exist? (Puppet::Environments::EnvironmentNotFound)
pdk (DEBUG): Execution of 'bundle.bat exec puppet debugger --basemodulepath=C:/Source/tmp/pdkmodule/spec/fixtures/modules:C:/Source/tmp/pdkmodule/modules --environmentpath=C:/Source/tmp/pdkmodule/spec/fixtures/modules:C:/Source/tmp/pdkmodule/modules --modulepath=C:/Source/tmp/pdkmodule/spec/fixtures/modules:C:/Source/tmp/pdkmodule/modules' complete (duration:                               16.243607s; exit code: 1)
C:\Source\tmp\pdkmodule>
```

glennsarti commented:
I don't have rights on your branch but here are my changes -

https://github.com/nwops/pdk/compare/console...glennsarti:corey-console

logicminds commented:
@glennsarti added your changes and updated my tests.

I also added a new check to ensure the user has fixtures downloaded.    Currently this is just a warning message as it only impacts the modules the debugger has access too.


```
pdk console
pdk (WARN): Module fixtures not found, please run pdk bundle exec rake spec_prep.
```

glennsarti commented:
Confirmed changes at commit 2be8a94 work as expected on Windows and Linux. As it's experimental edge cases can be dealt with later.

@rodjek Unless you have any other thoughts I'll merge soon.

glennsarti commented:
🎉🎉🎉🎉
